README.md
# Storage

![license](https://img.shields.io/badge/license-MIT-brightGreen.svg)
[![build](https://github.com/originphp/storage/workflows/CI/badge.svg)](https://github.com/originphp/storage/actions)
[![coverage](https://coveralls.io/repos/github/originphp/storage/badge.svg?branch=master)](https://coveralls.io/github/originphp/storage?branch=master)

The Storage library provides an easy way to access different types of storages from local disk, ZIP archives, FTP and SFTP. Its a unified approach for working with different storages.

## Installation

To install this package

```linux
$ composer require originphp/storage
```

## Configuration

You need to configure the default storage engine, you can use multiple engines, but there must be a default one.

In your bootstrap/configuration files add

```php
use Origin\Storage\Storage;
use Origin\Storage\Engine\LocalEngine;

Storage::config('default', [
    'className' => LocalEngine::class
    'root' => '/var/www/storage'
]);
```

## Using Storage

### Writing To Storage

```php
use Origin\Storage\Storage;
Storage::write('test.txt','hello world!');
```

You can also write to folders directly. Folders in the tree that do not exist will be created automatically.

```php
Storage::write('my_folder/test.txt','hello world!');
```

### Reading From Storage

```php
use Origin\Storage\Storage;
$contents = Storage::read('my_folder/test.txt');
```

### Deleting From Storage

To delete files or folders

```php
Storage::delete('my_folder/test.txt');
Storage::delete('my_folder');
```

Folders are deleted recursively automatically, when using delete.

### Listing Storage Contents

> Version 2.0 no longer lists names of files relative to the path you pass to the list method. The full path name is always returned.

To list the files on the storage

```php
use Origin\Storage\Storage;
$allFiles = Storage::list();
```

Storage contents are listed recursively and it will provide you with an array of `FileObjects`. Each file has is an object which can be accessed as an array or an object

```php
// Will look like this
Origin\Storage\FileObject Object
(
    [name] => bar.txt
    [directory] => folder
    [path] => folder/bar.txt
    [extension] => txt
    [timestamp] => 1601121922
    [size] => 32
)

echo $file->name;
echo $file['name'];
```

When the `FileObject` is converted to a string it will become a path e.g. `/main/subfolder/foo.txt`

If you just want the files of particular folder, then it will list all files recursively under that folder.

```php
use Origin\Storage\Storage;
$files = Storage::list('my_folder');
```

### Working with Multiple Storages


Whether you are using multiple storage engines, or you multiple configurations for a single storage engine, the Storage utility is flexible.

You can get the configured Storage volume

```php
$volume = Storage::volume('sftp-backup');
$data = $volume->read('transactions.csv');
```

Or you can pass an options array telling the Storage object which configuration to use

```php
$data = Storage::read('transactions.csv',[
    'config'=>'sftp-backup'
]);
```

## Storage Engines

### Local

The local storage simply works with data from the drive.

```php
use Origin\Storage\Storage;
use Origin\Storage\Engine\LocalEngine;

Storage::config('default', [
    'className' => LocalEngine::class
    'root' => '/var/www/storage',
    'lock' => true // default
 ]);
```

### FTP

Then you need to configure this

```php
use Origin\Storage\Storage;
use Origin\Storage\Engine\FtpEngine;

Storage::config('default', [
    'className' => FtpEngine::class
    'engine' => 'Ftp',
    'host' => 'example.com',
    'port' => 21,
    'username' => 'james',
    'password' => 'secret',
    'ssl' => false
 ]);
```

options for configuring FTP include:

- host: the hostname or ip address
- port: the port number. default 21
- username: the ftp username
- password: the ftp password
- timeout: default 10 seconds
- passive: deafult false
- root: the root folder of the storage within your ftp account
- ssl: default: false

### SFTP

To use the SFTP engine, you need to install `phpseclib`

```linux
$ composer require phpseclib/phpseclib:~2.0
```

Then configure as follows:

```php
use Origin\Storage\Storage;
use Origin\Storage\Engine\SftpEngine;

Storage::config('default', [
    'className' => SftpEngine::class
    'host' => 'example.com',
    'port' => 22,
    'username' => 'james',
    'password' => 'secret'
 ]);
```

If you use want to use a private key to login, you can either provide the filename with the full path or the contents of the private key itself.


```php
use Origin\Storage\Storage;
use Origin\Storage\Engine\SftpEngine;

Storage::config('default', [
    'className' => SftpEngine::class
    'host' => 'example.com',
    'port' => 22,
    'username' => 'james',
    'privateKey' => '/var/www/config/id_rsa'
]);
```

If your private key requires a password then you can provide that as well. See the [How to setup SSH keys ](https://linuxize.com/post/how-to-set-up-ssh-keys-on-ubuntu-1804/) tutorial for more information.

options for configuring SFTP include:

- host: the hostname or ip address
- port: the port number. default 22
- username: the ssh account username
- password: the ssh account password
- timeout: default 10 seconds
- root: the root folder of the storage. e.g. /home/user/sub_folder
- privateKey: either the private key for the account or the filename where the private key can be loaded from


### S3

The S3 Engine works with [Amazon S3](https://aws.amazon.com/s3/) and any other object storage server which uses the S3 protocol, for example [minio](https://min.io/).

To use the S3 Engine, you need to install the Amazon AWS SDK

```linux
$ composer require aws/aws-sdk-php
```

Then you can configure the S3 engine like this

```php
use Origin\Storage\Storage;
use Origin\Storage\Engine\S3Engine;

Storage::config('default', [
    'className' => S3Engine::class
    'credentials' => [
        'key' => env('S3_KEY'), // * required
        'secret' => env('S3_SECRET'), // * required
    ],
    'region' => 'us-east-1', // * required
    'version' => 'latest',
    'endpoint' => env('S3_ENDPOINT'), // for S3 comptabile protocols
    'bucket' => env('S3_BUCKET') // * required
 ]);
```

Options for configuring the `S3` engine are:

- credentials: this is required and is an array with both `key` and `secret`
- region: The label for location of the server
- version: version setting
- endpoint: If you are not using Amazon S3. e.g. `http://127.0.0.1:9000`
- bucket: The name of the bucket, this is required and the bucket should exist.

#### Minio Server (S3)

To fire up your own minio server locally you can run the docker command

```linux
$ docker run -p 9000:9000 minio/minio server /data
```

You can access this also using your web browser at `http://127.0.0.1:9000`.

## Zip

To use the ZIP storage engine, provide the filename with a full path.

```php
use Origin\Storage\Storage;
use Origin\Storage\Engine\ZipEngine;

Storage::config('default', [
    'className' => ZipEngine::class
    'file' => '/var/www/backup.zip'
]);
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.1.0] - 2021-04-13

### Changed

- Adjusted to use Defer 3.x

## [3.0.0] - 2021-01-04

### Fixed

- Fixed issue with FTP using non jail

### Changed

- Changed minimum PHP version to 7.3
- Changed minimum PHPUnit to 9.2
- Changed defer to 2.0
- Changed configurable to 2.0
- Changed FTPEngine to throw StorageException

### Added

- Added NotFoundException for MemoryEngine
- Added S3Engine::errors for debugging AWS or S3Exceptions

## [2.0.2] - 2020-10-24

### Fixed

- Fixed missing Storage::write type declaration for value

## [2.0.0] - 2020-09-26

### Changed

- Changed `list` return data, `name` now represents the `filename`, and added `path` which includes the directory and filename. Use `path` instead of `name`. (Breaking Change)

## [1.4.0] - 2020-09-26

### Added

- Added FileObject::\_\_toString which returns the path e.g. /folder/example.png

## [1.3.1] - 2020-09-03

### Fixed

- Fixed clearstatcache arguments

## [1.3.0] - 2020-09-03

### Added

- Added better lock support using flock for `LocalEngine`, this can also be disabled through config.

### Changed

- Changed directory mask when creating from 744 to 755

## [1.2.1] - 2020-07-23

### Fixed

- Fixed operators

## [1.2.0] - 2020-07-07

### Added

- Added S3 Engine

## [1.1.0] - 2019-12-15

### Added

- Added Memory Engine
- Added Zip Engine

## [1.0.1] - 2019-10-31

### Changed

- Changed Storage::list file result can be accessed as both array and object (FileObject)

### Fixed

- Fixed tests

## [1.0.0] - 2019-10-25

This component has been decoupled from the [OriginPHP framework](https://www.originphp.com/).


src/Engine/S3Engine.php
<?php
/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright    Copyright (c) Jamiel Sharief
 * @link         https://www.originphp.com
 * @license      https://opensource.org/licenses/mit-license.php MIT License
 */
declare(strict_types=1);
namespace Origin\Storage\Engine;

use Aws\Result;
use Aws\S3\S3Client;
use InvalidArgumentException;
use Origin\Storage\FileObject;
use Aws\Exception\AwsException;
use Aws\S3\Exception\S3Exception;
use Origin\Storage\Exception\FileNotFoundException;
use Aws\S3\Exception\DeleteMultipleObjectsException;

/**
 * S3 Engine
 * To fire up a minio contianer
 * $ docker run -p 9000:9000 minio/minio server /data
 *
 */
class S3Engine extends BaseEngine
{
    /**
     * @var Aws\S3\S3Client $s3
     */
    private $s3;

    /**
     * @var string
     */
    private $bucket;

    private $errors = [];

    protected $defaultConfig = [
        'credentials' => [
            'key' => null,
            'secret' => null,
        ],
        'region' => 'main-rack',
        'version' => 'latest',
        'endpoint' => null, // for S3 comptabile protocols
        'bucket' => 'data'
    ];

    protected function initialize(array $config): void
    {
        $credentials = $this->config('credentials');

        if ($credentials === null || empty($credentials['key']) || empty($credentials['secret'])) {
            throw new InvalidArgumentException('Invalid Credentials settings');
        }
        if ($this->config('region') === null) {
            throw new InvalidArgumentException('Region not set');
        }
        if ($this->config('bucket') === null) {
            throw new InvalidArgumentException('Bucket not set');
        }

        $this->s3 = new S3Client($config);
        $this->bucket = $config['bucket'];
    }

    /**
     * Returns a list of buckets on the S3 storage
     * @internal can be slight delay even after creating before you can read
     *
     * @return array
     */
    public function listBuckets(): array
    {
        $out = [];
        $buckets = $this->s3->listBuckets();
        foreach ($buckets['Buckets'] as $bucket) {
            $out[] = $bucket['Name'];
        }

        return $out;
    }

    /**
     * Creates a bucket
     *
     * @param string $name
     * @return boolean
     */
    public function createBucket(string $name): bool
    {
        $options = ['Bucket' => $name];
    
        try {
            $this->s3->createBucket($options);
            $this->s3->waitUntil('BucketExists', $options);

            return true;
        } catch (AwsException $exception) {
            $this->errors[] = $exception->getMessage();
        }

        return false;
    }

    /**
     * Deletes a bucket
     *
     * @param string $name
     * @return boolean
     */
    public function deleteBucket(string $name): bool
    {
        $options = ['Bucket' => $name];

        try {
            $this->s3->deleteBucket($options);
            $this->s3->waitUntil('BucketNotExists', $options);

            return true;
        } catch (AwsException $exception) {
            $this->errors[] = $exception->getMessage();
        }

        return false;
    }

    /**
     * Reads from the storage
     *
     * @param string $name
     * @return string
     */
    public function read(string $name): string
    {
        try {
            $response = $this->s3->getObject([
                'Bucket' => $this->bucket,
                'Key' => trim($name, '/'),
            ]);

            return (string) $response['Body'];
        } catch (S3Exception $exception) {
            $this->errors[] = $exception->getMessage();
        }
    
        throw new FileNotFoundException(sprintf('%s does not exist', $name));
    }

    /**
     * Writes to the storage
     *
     * @see https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html
     *
     * @param string $name
     * @param string $data
     * @return bool
     */
    public function write(string $name, string $data): bool
    {
        try {
            $result = $this->s3->putObject([
                'Bucket' => $this->bucket,
                'Key' => trim($name, '/'),
                'Body' => $data
            ]);

            return true;
        } catch (S3Exception $exception) {
            $this->errors[] = $exception->getMessage();
        }

        return false;
    }

    /**
     * Deletes from the storage
     *
     * @param string $name
     * @return bool
     */
    public function delete(string $name): bool
    {
        // Disable anything with trailing /
        if (substr($name, -1) === '/') {
            return false;
        }

        if (! $this->exists($name)) {
            throw new FileNotFoundException(sprintf('%s does not exist', $name));
        }
        
        if ($this->isDirectory($name)) {
            return $this->deleteDirectory($name);
        }

        return $this->deleteObject($name);
    }

    /**
     * Deletes all items in directory
     *
     * @param string $name
     * @return boolean
     */
    private function deleteDirectory(string $name): bool
    {
        try {
            $this->s3->deleteMatchingObjects($this->bucket, trim($name, '/') . '/');
        } catch (DeleteMultipleObjectsException $exception) {
            $this->errors[] = $exception->getMessage();
            return false;
        }

        return true;
    }

    /**
     * @param string $name
     * @return boolean
     */
    private function deleteObject(string $name): bool
    {
        try {
            $result = $this->s3->deleteObject([
                'Bucket' => $this->bucket,
                'Key' => $name,
            ]);

            return empty($result['DeleteMarker']);
        } catch (S3Exception $exception) {
            $this->errors[] = $exception->getMessage();
        }

        return false;
    }

    /**
     * Checks if a file or directory exists on the storage
     *
     * @link https://docs.aws.amazon.com/AmazonS3/latest/dev/ListingObjectKeysUsingPHP.html
     * @link http://url.com https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-s3-2006-03-01.html#listobjects
     *
     * @param string $name
     * @return bool
     */
    public function exists(string $name): bool
    {
        $name = trim($name, '/');
        if ($this->s3->doesObjectExist($this->config('bucket'), $name)) {
            return true;
        }
        
        return $this->isDirectory($name);
    }

    /**
     * Returns the list of files from the storage
     *
     * @param string $name
     * @return array
     */
    public function list(string $name = null): array
    {
        $name = trim((string) $name, '/');

        if (! empty($name) && ! $this->isDirectory($name)) {
            throw new FileNotFoundException('directory does not exist');
        }
       
        $files = [];
        try {
            $results = $this->s3->getPaginator('ListObjects', [
                'Bucket' => $this->bucket,
                'Prefix' => $name
            ]);
            
            foreach ($results as $result) {
                foreach ($this->getListObjects($result) as $object) {
                    $info = $this->pathinfo($this->rebase($object['Key'], '/'));
                    $info['timestamp'] = strtotime((string) $object['LastModified']);
                    $info['size'] = $object['Size'];
                    $files[] = new FileObject($info);
                }
            }
        } catch (S3Exception $exception) {
            $this->errors[] = $exception->getMessage();
        }

        return $files;
    }

    /**
     * Undocumented function
     *
     * @param \Aws\Result $result
     * @return array
     */
    private function getListObjects(Result $result): array
    {
        $contents = $result->get('Contents') ?? [];
        $common = $result->get('CommonPrefixes') ?? [];

        return array_merge($contents, $common);
    }

    /**
     * Checks if an object is a directory
     *
     * @param string $name
     * @return boolean
     */
    private function isDirectory(string $name): bool
    {
        $command = $this->s3->getCommand('listObjects', [
            'Bucket' => $this->bucket,
            'Prefix' => $name . '/',
            'MaxKeys' => 1,
        ]);

        try {
            $result = $this->s3->execute($command);

            return ! empty($result['Contents'] || ! empty($result['CommonPrefixes']));
        } catch (S3Exception $exception) {
            $this->errors[] = $exception->getMessage();
        }

        return false;
    }

    /**
     * Gets the internal errors thrown by AWS, this can be
     * handy to debug problems.
     *
     * @return array
     */
    public function errors() : array
    {
        return $this->errors;
    }
}
